### PR TITLE
Deng 2219 telemetry dev cycle glean metrics

### PR DESCRIPTION
**Pull request by lelilia**
[ticket](https://mozilla-hub.atlassian.net/browse/DENG-2219)

cleaning up the glean metrics data and calculating the real expiration date from the last date data was sent and the expires column

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-2240)

### CODE REVIEW

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+description: |-
+  Dates for major firefox releases in release, beta and nightly
+owners:
+- leli@mozilla.com
+labels:
+  incremental: false
+  owner1: leli@mozilla.com
+scheduling:
+  dag_name: bqetl_telemetry_dev_cycle
+bigquery: null
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/metadata.yaml:11]:
```suggestion
```

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/metadata.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+references: {}
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/metadata.yaml:12]:
```suggestion
```

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql [hunk 3]
```diff
@@ -0,0 +1,17 @@
+WITH firefox_releases AS (
+  SELECT
+    CAST(REGEXP_EXTRACT(build.target.version, r"^\d+") AS INT64) AS version,
+    build.target.channel,
+    DATE(MIN(build.download.date)) AS publish_date
+  FROM
+    `mozdata.telemetry.buildhub2`
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql:7]:
```suggestion
    telemetry.buildhub2
```

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  WHERE
+    build.target.channel IN ("release", "beta", "nightly")
+  GROUP BY
+    version,
+    channel
+)
+SELECT
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql:14]: Seems like unnecessary stage here? We could just remove the CTE syntax and not have the SELECT * here?

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql:1]: Just curious, how does this compare to: https://github.com/mozilla/bigquery-etl/blob/main/sql/moz-fx-data-shared-prod/telemetry_derived/releases_v1/query.py ?

Are they overlapping? Is the other model still valid?

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    glean.type,
+    glean.first_seen_date AS release_date,
+    glean.last_seen_date AS last_date,
+    glean.expires,
+    CASE
+      WHEN glean.expires = "never"
+        THEN NULL
+      WHEN REGEXP_CONTAINS(glean.expires, r'[0-9]{4}-[0-9]{2}-[0-9]{2}')
+        THEN CAST(glean.expires AS DATE)
+      ELSE releases.publish_date
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:14]: Should the publish date be set to publish date if no expiry date set? Sounds like potentially this should be a `null` value?

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    END AS expiry_date,
+  FROM
+    `telemetry_dev_cycle_external.glean_metrics_stats_v1` AS glean
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:17]:
```suggestion
    telemetry_dev_cycle_external.glean_metrics_stats_v1 AS glean
```

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  LEFT JOIN
+    `telemetry_dev_cycle_derived.firefox_major_release_dates_v1` AS releases
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:19]:
```suggestion
    telemetry_dev_cycle_derived.firefox_major_release_dates_v1 AS releases
```

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  ON
+    glean.expires = CAST(releases.version AS STRING)
+    AND COALESCE(
+      REGEXP_EXTRACT(glean.glean_app, r"beta"),
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:23]: Is this just checking if it contains a specific substring?

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    last_date,
+    expires,
+    CASE
+      WHEN expiry_date IS NULL
+        AND DATE_DIFF(CURRENT_DATE(), last_date, day) < 2
+        THEN NULL
+      WHEN expiry_date IS NULL
+        THEN last_date
+      ELSE IF(expiry_date < last_date, expiry_date, last_date)
+    END AS expired_date
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:43]: Should we have a default here with an `ELSE`?

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  FROM
+    glean_app_with_parsed_expiry_date
+)
+SELECT
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:47]: Seems like an unnecessary layer here. I'd just make final the SQL statement without wrapping it into a CTE.

**Review by kik-kik — APPROVED** [review]
Some minor comments so r+wc.

One thing I'd also suggest to think about is if adding any data checks to those queries would make sense here.

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql (hunk 3 above)
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql:7]: but you can't leave the mozdata out .. it's running in `moz-fx-data-shared-prod`?

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql (hunk 4 above)
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql:14]: this is left over from my dbt days ... where you always do a select * from final at the end ... so much easier to debug ... and the data is small enough that it's not a problem to select it again?

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql:1]: they don't have nightly releases in there and I wanted them

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql (hunk 5 above)
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:14]: null value indicates that it did not expire yet ...

Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:14]: so what's happening here is `expires` can either be "never" (-> null) a date (-> the date) or a version number. If it's a version number we take the date of the version being published as the date it expired

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql (hunk 6 above)
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:17]: why do you don't like the backticks?

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql (hunk 8 above)
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:23]: no this is in the join 
so if the `expires` is a version number, i need to join it with the releases table on the version and the channel. if there is a beta in the glean_app, I'm joining it to beta channel, if there is a "nightly" to nightly and else to release

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql (hunk 9 above)
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/glean_metrics_stats_v1/query.sql:43]: what do you mean by that?

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql (hunk 3 above)
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql:7]: should also be in moz-prod

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql (hunk 4 above)
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql:14]: follow up ticket

File: sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql
Comment (lelilia) [reply sql/moz-fx-data-shared-prod/telemetry_dev_cycle_derived/firefox_major_release_dates_v1/query.sql:1]: follow up ticket

**Pull request merged**